### ISSUE TITLE
How to mock instance field with static method ?

### ISSUE THREAD

**Issue by oddstab** [Serg046/AutoFake#211]
I have this tests:
```csharp
public class TestClass
{
    public int Field = StaticTestClass.Get111();
    //public int Field = 111;

    public int Something()
    {
        return Field;
    }
}

public class StaticTestClass
{
    public static int Get111()
    {
        throw new Exception("QAQ");
    }
}

[TestClass]
public class Question
{
    [TestMethod]
    public void Get222Test()
    {
        var fake = new Fake<TestClass>();

        var sut = fake.Rewrite(f => f.Something());
        sut.Replace(s => s.Field).Return(222);

        Assert.AreEqual(222, sut.Execute());
    }
}
```
Test will fail, but when uncomment ```public int Field = 111;``` test will pass, WHY?

**Comment by Serg046** [comment]
Looks like a bug. The following codfe prints `222` as expected, not `111`
```
class Program
{
	static void Main(string[] args)
	{
		var fake = new Fake<TestClass>();

		var sut = fake.Rewrite(f => f.Something());
		sut.Replace(s => s.Field).Return(222);

		Console.WriteLine(sut.Execute());
	}
}

public class TestClass
{
	public int Field = StaticTestClass.Get111();

	public int Something()
	{
		return Field;
	}
}

public class StaticTestClass
{
	public static int Get111()
	{
		return 111;
	}
}
```

**Comment by oddstab** [comment]
> Not a bug, your `TestClass` will be actually compiled to:
> 
> ```
> public class TestClass
> {
>     public int Field;
>     
>     public TestClass()
>     {
>         Field = StaticTestClass.Get111(); // It just throws here even before AutoFake processes the method
>     }
> 
>     public int Something()
>     {
>         return Field;
>     }
> }
> ```
> 
> So you actually want to tune the constructor, not the method:
> 
> ```
> class Program
> {
> 	static void Main(string[] args)
> 	{
> 		var fake = new Fake<TestClass>();
> 
> 		fake.Rewrite(() => new TestClass()).Replace(() => StaticTestClass.Get111()).Return(222);
> 
> 		Console.WriteLine(fake.Execute(f => f.Something()));
> 	}
> }
> 
> public class TestClass
> {
> 	public int Field = StaticTestClass.Get111();
> 
> 	public int Something()
> 	{
> 		return Field;
> 	}
> }
> 
> public class StaticTestClass
> {
> 	public static int Get111()
> 	{
> 		throw new Exception("QAQ");
> 	}
> }
> ```
> 
> But there was another issue with this updated code sample above related to mocking in ctors. It is now fixed in the latest version v0.4.3.7. Could you review the answer and see if we can close this issue?
> 
> [9f1698b](https://github.com/Serg046/AutoFake/commit/9f1698baba027c2cb787b6a0f156a891c8d95fea)

seems to work, but when change to this

```csharp
public class TestClass
{
    public int Something()
    {
        var t = new ThirdClass();
        return t.Field;
    }
}

public class StaticTestClass
{
    public static int Get111()
    {
        throw new Exception("QAQ");
    }
}

public class ThirdClass
{
    public int Field = StaticTestClass.Get111();
}
```

write some tests
```csharp
[TestMethod]
public void MyTestMethod1()
{
    var fake = new Fake<TestClass>();

    fake.Rewrite(() => new ThirdClass())
        .Replace(() => StaticTestClass.Get111())
        //.Replace((ThirdClass c) => c.Field)
        .Return(222);

    int result = fake.Execute(c => c.Something());
    Assert.AreEqual(222, result);
}

[TestMethod]
public void MyTestMethod2()
{
    var fake = new Fake<TestClass>();

    var sut = fake.Rewrite((c) => c.Something());

    sut.Replace(() => new ThirdClass()).Return(new ThirdClass { Field = 222 });

    int result = sut.Execute();
    Assert.AreEqual(222, result);
}
```

still got error, even uncomment `//.Replace((ThirdClass c) => c.Field)`

**Comment by oddstab** [comment]
> @oddstab, your first test should rewrite the original "entrypoint" method and the second test fails because `new ThirdClass { Field = 222 }` fails, see the following examples:
> 
> ```
> class Program
> {
> 	static void Main(string[] args)
> 	{
> 		var fake1 = new Fake<TestClass>();
> 		var sut1 = fake1.Rewrite(f => f.Something()); // You should rewrite the entry point
> 		sut1.Replace(() => StaticTestClass.Get111()).Return(222);
> 		Console.WriteLine(sut1.Execute());
> 
> 		try
> 		{
> 			var fake2 = new Fake<TestClass>();
> 			var sut2 = fake2.Rewrite(c => c.Something());
> 			sut2.Replace(() => new ThirdClass()).Return(new ThirdClass {Field = 222});
> 			Console.WriteLine(sut1.Execute());
> 		}
> 		catch
> 		{
> 			Console.WriteLine("This is correct, just run \"new ThirdClass {Field = 222}\", it fails because of its constructor");
> 		}
> 
> 		// This is the proof for the example above
> 		var fake3 = new Fake<TestClass>();
> 		var sut3 = fake3.Rewrite(c => c.Something());
> 		sut3.Replace(() => new ThirdClass2()).Return(new ThirdClass2 { Field = 222 });
> 		Console.WriteLine(sut1.Execute());
> 	}
> }
> 
> public class TestClass
> {
> 	public int Something()
> 	{
> 		var t = new ThirdClass();
> 		return t.Field;
> 	}
> 
> 	public int Something2()
> 	{
> 		var t = new ThirdClass2();
> 		return t.Field;
> 	}
> }
> 
> public class StaticTestClass
> {
> 	public static int Get111()
> 	{
> 		throw new Exception("QAQ");
> 	}
> }
> 
> public class ThirdClass
> {
> 	public int Field = StaticTestClass.Get111();
> }
> 
> public class ThirdClass2
> {
> 	public int Field { get; set; }
> }
> ```
> 
> Output:
> 
> ```
> 222
> This is correct, just run "new ThirdClass {Field = 222}", it fails because of its constructor
> 222
> ```

Thank you. it is working.
